File: src/org/jgll/recognizer/AbstractGLLRecognizer.java
package org.jgll.recognizer;

import java.util.Deque;
import java.util.Map;
import java.util.Set;

import org.jgll.grammar.Grammar;
import org.jgll.grammar.GrammarSlot;
import org.jgll.grammar.HeadGrammarSlot;
import org.jgll.util.Input;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public abstract class AbstractGLLRecognizer implements GLLRecognizer {
	
	private static final Logger log = LoggerFactory.getLogger(AbstractGLLRecognizer.class);
	
	protected Input input;
	
	/**
	 * The current input index.
	 */
	protected int ci;

	/**
	 * The current GSS node.
	 */
	protected GSSNode cu;

	/**
	 * u0 is the bottom of the GSS.
	 */
	protected GSSNode u0;
	
	/**
	 * 
	 */
	protected Grammar grammar;
	
	/**
	 * The nonterminal from which the parsing will be started.
	 */
	protected HeadGrammarSlot startSymbol;
	
	private Set<Descriptor> descriptorSet;
	
	private Deque<Descriptor> descriptorStack;
	
	private Map<GSSNode, GSSNode> gssNodes;
	
	@Override
	public final boolean recognize(Input input, Grammar grammar, String nonterminalName) {		
		HeadGrammarSlot startSymbol = grammar.getNonterminalByName(nonterminalName);
		if(startSymbol == null) {
			throw new RuntimeException("No nonterminal named " + nonterminalName + " found");
		}
		
		this.input = input;
		this.grammar = grammar;
		this.startSymbol = startSymbol;
		
		init();
	
		long start = System.nanoTime();
	
		recognize(startSymbol);
		
		long end = System.nanoTime();

		logStatistics(end - start);

		if(descriptorSet.contains(new Descriptor(startSymbol, u0, input.size()))) {
			return true;
		} else {
			return false;
		}

	}
	
	private void logStatistics(long duration) {
		log.info("Parsing Time: {} ms", duration/1000000);
		
		int mb = 1024 * 1024;
		Runtime runtime = Runtime.getRuntime();
		log.debug("Memory used: {} mb", (runtime.totalMemory() - runtime.freeMemory()) / mb);
//		log.debug("Descriptors: {}", lookupTable.getDescriptorsCount());
//		log.debug("GSSNodes: {}", lookupTable.getGSSNodes().size());
//		log.debug("Non-packed nodes: {}", lookupTable.getDescriptorsCount());
//		log.debug("GSS Nodes: {}", lookupTable.getGSSNodesCount());
//		log.debug("GSS Edges: {}", lookupTable.getGSSEdgesCount());
	}

	
	protected abstract void recognize(HeadGrammarSlot startSymbol);
	
	/**
	 * initialized the parser's state before a new parse.
	 */
	protected abstract void init();

	@Override
	public final void add(GrammarSlot label, GSSNode u, int inputIndex) {
		Descriptor descriptor = new Descriptor(label, u, inputIndex);
		if(descriptorSet.add(descriptor)) {
			log.trace("Descriptor added: {}", descriptor);
			descriptorStack.push(descriptor);
		} else {
			log.trace("Descriptor {} already exists.", descriptor);
		}
	}
	
	@Override
	public final void pop(GSSNode u, int i) {
		
		log.trace("Pop {}, {}, {}", u.getLabel(), i);
		
		if (u != u0) {
			
			// Add (u, i) to P
			u.addPoppedIndex(i);
			
			for(GSSNode node : u.getChildren()) {
				add(u.getLabel(), node, i);
			}			
		}
	}

	@Override
	public final GSSNode create(GrammarSlot L, GSSNode u, int i) {
		log.trace("GSSNode created: " +  L + ", " + i);
		GSSNode key = new GSSNode(L, i);

		GSSNode v = gssNodes.get(key);
		if(v == null) {
			v = key;
			gssNodes.put(key, v);
			return key;
		}
		
		for(int index : v.getPoppedIndices()) {
			add(L, u, index);
		}
		
		return v;
	}
	
	@Override
	public boolean hasNextDescriptor() {
		return !descriptorStack.isEmpty();
	}
	
	@Override
	public Descriptor nextDescriptor() {
		return descriptorStack.pop();
	}

}

File: src-scala/org/jgll/recognizer/AbstractGLLRecognizer.scala
package org.jgll.recognizer

import java.util.ArrayDeque
import java.util.Deque
import org.jgll.grammar.Grammar
import org.jgll.grammar.slot.BodyGrammarSlot
import org.jgll.grammar.slot.GrammarSlot
import org.jgll.grammar.slot.HeadGrammarSlot
import org.jgll.grammar.slot.L0
import org.jgll.grammar.slot.StartSlot
import org.jgll.util.hashing.CuckooHashSet
import org.jgll.util.logging.LoggerWrapper
import org.jgll.util.InputTrait

import AbstractGLLRecognizer._
object AbstractGLLRecognizer {

  private val log = LoggerWrapper.getLogger(classOf[AbstractGLLRecognizer])

  val startSlot = new StartSlot("Start")

  val u0 = GSSNode.U0
}

abstract class AbstractGLLRecognizer extends GLLRecognizer {

  protected var input: Input = _

  protected var ci: Int = _

  protected var cu: GSSNode = u0

  protected var grammar: Grammar = _

  protected var startSymbol: HeadGrammarSlot = _

  protected var descriptorSet: CuckooHashSet[Descriptor] = _

  protected var descriptorStack: Deque[Descriptor] = _

  protected var gssNodes: CuckooHashSet[GSSNode] = _

  protected var recognized: Boolean = _

  protected var endIndex: Rep[Int] = _

  override def recognize(input: Input, grammar: Grammar, nonterminalName: String): Boolean = {
    val startSymbol = grammar.getNonterminalByName(nonterminalName)
    if (startSymbol == null) {
      throw new RuntimeException("No nonterminal named " + nonterminalName + " found")
    }
    init(grammar, input, 0, input.size - 1, startSymbol)
    cu = create(startSlot, cu, ci)
    val start = System.nanoTime()
    L0.getInstance.recognize(this, input, startSymbol)
    val end = System.nanoTime()
    logStatistics(end - start)
    recognized
  }

  override def recognize(input: Input,
      startIndex: Int,
      endIndex: Rep[Int],
      fromSlot: BodyGrammarSlot): Boolean = {
    init(grammar, input, startIndex, endIndex, null)
    cu = create(startSlot, cu, ci)
    val start = System.nanoTime()
    add(fromSlot, cu, ci)
    L0.getInstance.recognize(this, input)
    val end = System.nanoTime()
    logStatistics(end - start)
    recognized
  }

  protected def logStatistics(duration: Long) {
    log.debug("Recognition Time: %d ms", duration / 1000000)
    val mb = 1024 * 1024
    val runtime = Runtime.getRuntime
    log.debug("Memory used: %d mb", (runtime.totalMemory() - runtime.freeMemory()) / mb)
  }

  protected def init(grammar: Grammar,
      input: Input,
      startIndex: Int,
      endIndex: Rep[Int],
      startSymbol: HeadGrammarSlot) {
    this.grammar = grammar
    this.startSymbol = startSymbol
    this.input = input
    this.ci = startIndex
    this.endIndex = endIndex
    this.recognized = false
    this.cu = u0
    if (descriptorSet == null) {
      descriptorSet = new CuckooHashSet(Descriptor.externalHasher)
    } else {
      descriptorSet.clear()
    }
    if (descriptorStack == null) {
      descriptorStack = new ArrayDeque()
    } else {
      descriptorStack.clear()
    }
    if (gssNodes == null) {
      gssNodes = new CuckooHashSet(GSSNode.externalHasher)
    } else {
      gssNodes.clear()
    }
  }

  override def add(slot: GrammarSlot, u: GSSNode, inputIndex: Int) {
    val descriptor = new Descriptor(slot, u, inputIndex)
    if (descriptorSet.add(descriptor) == null) {
      log.trace("Descriptor added: %s : true", descriptor)
      descriptorStack.push(descriptor)
    } else {
      log.trace("Descriptor %s : false", descriptor)
    }
  }

  override def pop(u: GSSNode, i: Int) {
    log.trace("Pop %s, %d", u.getGrammarSlot, i)
    if (u != u0) {
      u.addPoppedIndex(i)
      for (node <- u.getChildren) {
        add(u.getGrammarSlot, node, i)
      }
    }
  }

  override def create(L: GrammarSlot, u: GSSNode, i: Int): GSSNode = {
    log.trace("GSSNode created: (%s, %d)", L, i)
    val key = new GSSNode(L, i)
    var v = gssNodes.add(key)
    if (v == null) {
      v = key
    }
    if (!v.hasChild(u)) {
      v.addChild(u)
      for (index <- v.getPoppedIndices) {
        add(L, u, index)
      }
    }
    v
  }

  override def hasNextDescriptor(): Boolean = !descriptorStack.isEmpty

  override def nextDescriptor(): Descriptor = descriptorStack.pop()

  override def update(gssNode: GSSNode, inputIndex: Int) {
    this.ci = inputIndex
    this.cu = gssNode
  }

  override def recognitionError(gssNode: GSSNode, inputIndex: Int) {
  }

  override def getCi(): Int = ci

  override def getCu(): GSSNode = cu
}
